### ISSUE TITLE
Add EllipsoidalNeighborhood

### ISSUE THREAD

**Issue by rmcaixeta** [JuliaEarth/GeoStats.jl#144]
After #140, now I think it's possible to do an ellipsoidal search using `KBallSearch` or `KNearestSearch` together with an `aniso2distance` metric

The proposal here is to create an `EllipsoidalNeighborhood` search with more constraints, similar to the ones used in mining softwares. The neighbors are initially restricted to the given ellipsoid. And the valid neighbors are selected considering some restrictions: `maxneigh`, `maxperkey` (e.g. max samples from a same drill hole) and `maxperoct` (max samples per ellipsoid octant). The priority is given to the nearest samples considering given metric (euclidean, aniso2distance, etc). Example:

```julia
ranges = [60.0, 30.0, 5.0]
rotangles = [30, 30, 20]
e = EllipsoidalNeighborhood(ranges, rotangles, convention=:GSLIB, maxperoct=3, maxperkey=(:holeid,3), metric=Euclidean())
s = NeighborhoodSearch(georefdata, e)
```

Let me know if you have some suggestions

**Comment by juliohm** [comment]
That would be amazing. A `EllipseNeighborhood` could use the `aniso2distance` helper function in the constructor to save the appropriate `Mahalanobis` distance as a field of the struct, and then a new search method with additional constraints could filter the results as you mentioned. 💯

Feel free to assign yourself to the issue if you would like to work on it. :)

**Comment by juliohm** [comment]
I think we don't even need a `EllipseNeighborhood`, we can simply use a `BallNeighborhood` with the `Mahalanobis` produced by `aniso2distance` 👍🏽

**Comment by juliohm** [comment]
So maybe, we just need to adapt the existing `NeighbordhooSearch` method to post-process the encountered neighbors.

**Comment by rmcaixeta** [comment]
You mean just extending the `BallNeighborhood` with `maxperkey` and `maxperoct`?
An do the neighbours post-processing inside `search` function?

**Comment by juliohm** [comment]
Maybe the search method should include the options `maxperkey` and `maxperoct`. So the `BallNeighborhood` stays untouched and is just used internally in the search method with post-processing steps. Makes sense?

**Comment by juliohm** [comment]
I think we can concentrate on the options `maxperkey` and `maxperoct` in some existing search method if that is possible, and then later consider writing some helper function to construct the appropriate `BallNeighborhood` from `aniso2distance`.

**Comment by rmcaixeta** [comment]
Ok, I'll try that. I have something already written here, just need to adapt and test

**Comment by juliohm** [comment]
In terms of design, we could even consider first creating a separate `Ellipse` geometry to hold the semiaxes and rotations as a standalone object, and then refactor `aniso2distance` to take this `Ellipse` and convert to a distance. But we can handle this refactoring in a separate PR.

**Comment by juliohm** [comment]
At some point, we could have:

1. `Ellipse` geometry given `semiaxes` and `rotation` specs.
2. `distance(::Ellipse)` that returns the corresponding `Mahalanobis`
3. `maxperkey` and `maxperoct` keyword options in some existing search method
4. `EllipseNeighborhood` constructor that returns a `BallNeighborhood` with correct metric

We can try to address these issues step-by-step after your PR is merged.

**Comment by juliohm** [comment]
Updated the comment to reflect some reconsiderations.

**Comment by rmcaixeta** [comment]
Ok. I liked the previous idea too
By including `maxperkey` and `maxperoct` you are thinking something similar to that?

```julia
function search(xₒ::AbstractVector, method::NeighborhoodSearch{O,N,K};
  mask=nothing, maxperkey, maxperoct, maxneigh) where {O,N<:BallNeighborhood,K}
```

How to avoid ambiguities and method conflicts with `search(xₒ::AbstractVector, method::NeighborhoodSearch{O,N,T}; mask=nothing) where {O,N<:BallNeighborhood,T}` in that case?

**Comment by juliohm** [comment]
No, I am thinking that the method specification should include these options:

```julia
struct NeighborhoodSearch
  ...
  maxperkey
  maxperoct
end
```

The `search` methods stay untouched to be consistent with other search methods.

**Comment by juliohm** [comment]
Maybe start with a PR that just adds the `EllipseNeighborhood` to fix this issue here, and then we can open another issue and PR to figure out where these search options should go, if on an existing method or a new search method.

**Comment by juliohm** [comment]
Thanks @rmcaixeta for this great contribution. Closing it as it is now fixed on master 💯
